File: libs/power-log-parser/src/lib/parsers/battlegrounds-hero-reroll-parser.ts
import { GameTag } from '@firestone-hs/reference-data';
import { ActionParser } from '../action-parser';
import { GameEventHelper, GameEventProvider } from '../game-event';
import { ChangeEntity, Node, NodeType } from '../models';
import { GameState } from '../state/game-state';
import { ParserState, StateType } from '../state/parser-state';
import type { StateFacade } from '../state/state-facade';

export class BattlegroundsHeroRerollParser implements ActionParser {
	readonly ParserName = 'BattlegroundsHeroRerollParser';

	private GameState: GameState;
	private ParserState: ParserState;
	private StateFacade: StateFacade;

	constructor(parserState: ParserState, helper: StateFacade) {
		this.ParserState = parserState;
		this.GameState = parserState.GameState;
		this.StateFacade = helper;
	}

	AppliesOnNewNode(_node: Node, _stateType: StateType): boolean {
		return false;
	}

	AppliesOnCloseNode(node: Node, stateType: StateType): boolean {
		return (
			stateType === StateType.PowerTaskList &&
			this.StateFacade.IsBattlegrounds() &&
			node.Type === NodeType.ChangeEntity &&
			(node.Object as ChangeEntity).GetTag(GameTag.BACON_NUM_MULLIGAN_REFRESH_USED) > 0
		);
	}

	CreateGameEventProviderFromNew(_node: Node): GameEventProvider[] | null {
		return null;
	}

	CreateGameEventProviderFromClose(node: Node): GameEventProvider[] | null {
		const changeEntity = node.Object as ChangeEntity;
		return [
			GameEventProvider.Create(
				changeEntity.TimeStamp,
				'BATTLEGROUNDS_HERO_REROLL',
				GameEventHelper.CreateProvider(
					'BATTLEGROUNDS_HERO_REROLL',
					changeEntity.CardId,
					-1,
					changeEntity.Entity,
					this.StateFacade,
					null,
				),
				true,
				node,
			),
		];
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Parsers/BattlegroundsHeroRerollParser.cs
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using System.Linq;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using System.Collections.Generic;
using static HearthstoneReplays.Events.CardIds;

namespace HearthstoneReplays.Events.Parsers
{
    public class BattlegroundsHeroRerollParser : ActionParser
    {
        private GameState GameState { get; set; }
        private ParserState ParserState { get; set; }
        private StateFacade StateFacade { get; set; }

        public BattlegroundsHeroRerollParser(ParserState ParserState, StateFacade helper)
        {
            this.ParserState = ParserState;
            this.GameState = ParserState.GameState;
            this.StateFacade = helper;
        }

        public bool AppliesOnNewNode(Node node, StateType stateType)
        {
            return false;
        }

        public bool AppliesOnCloseNode(Node node, StateType stateType)
        {
            return stateType == StateType.PowerTaskList
                && StateFacade.IsBattlegrounds()
                && node.Type == typeof(ChangeEntity)
                && (node.Object as ChangeEntity).GetTag(GameTag.BACON_NUM_MULLIGAN_REFRESH_USED) > 0;
        }

        public List<GameEventProvider> CreateGameEventProviderFromNew(Node node)
        {
            return null;
        }

        public List<GameEventProvider> CreateGameEventProviderFromClose(Node node)
        {
            var changeEntity = node.Object as ChangeEntity;
            return new List<GameEventProvider> { GameEventProvider.Create(
                changeEntity.TimeStamp,
                "BATTLEGROUNDS_HERO_REROLL",
                GameEvent.CreateProvider(
                    "BATTLEGROUNDS_HERO_REROLL",
                    changeEntity.CardId,
                    -1,
                    changeEntity.Entity,
                    StateFacade,
                    null),
                true,
                node)
            };
        }
    }
}
